### PR TITLE
arch: arc: update arc connect debug select cpu mask when new cpu come online

### PR DESCRIPTION
**Pull request by IRISZZW**
The ARConnect Inter-core Debug Unit (ICD) provides additional
debug assist features in multi-core scenarios. It's useful to halt
other cores when one core is halted.

Before we program ICD in master core(core 0) initial stage, add
all cores to mask. so we need to make sure other slave cores have
launched and in running mode before we enable ICD in master core.

If we launch master core first, then launch slave cores by master core
conditionally, in this scenario, it's not OK.

Let's update arc connect debug (ARConnect ICD) select mask when
new slave core come online by slave core self, instead of use hardcoded
select mask.

### CODE REVIEW

**Review by YuguoWH — APPROVED** [review]

**Comment by ruuddw** [comment]
Great, this looks like a much more robust solution than #36535.
Can you please add a description of the problem this PR solves (like in the other PR), or even better create an issue for that?

**Review by abrodkin — CHANGES_REQUESTED** [review]
While implementation LGTM I'd like to see a rationale for this change clearly stated in the commit message. I.e. why do we want this way and why it's better than what we used to have.

**Comment by evgeniy-paltsev** [comment]
Should we revert #36535 (it has already been merged)?

File: arch/arc/core/arc_smp.c [hunk 1]
```diff
@@ -66,17 +66,40 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
...
 	z_irq_setup();
 
 	z_arc_connect_ici_clear();
 	z_irq_priority_set(IRQ_ICI, ARCV2_ICI_IRQ_PRIORITY, 0);
 	irq_enable(IRQ_ICI);
+
+	if (bcr.dbg) {
+	/* configure inter-core debug unit if available */
+		arc_connect_debug_setup(cpu_num);
+	}
```
Comment (evgeniy-paltsev) [inline arch/arc/core/arc_smp.c:102]: I'd place this right after we read BCR, so we update the debug mask as early as possible.

File: arch/arc/core/arc_smp.c [hunk 2]
```diff
@@ -66,17 +66,40 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
 	}
 }
 
+#ifdef CONFIG_SMP
+static void arc_connect_debug_setup(int cpu_num)
```
Comment (evgeniy-paltsev) [inline arch/arc/core/arc_smp.c:70]: Just a nitpick - probably name it something like `arc_connect_debug_mask_update` so something similar as we call it multiple times.

**Comment by ruuddw** [comment]
> Should we revert #36535 (it has already been merged)?

I thought about adding a DNM, but didn't do that since the PRs are orthogonal and reversing the order doesn't hurt.

**Comment by abrodkin** [comment]
@evgeniy-paltsev agree with @ruuddw - these are orthogonal changes and start of the master in the very end makes sense IMHO anyways.

File: arch/arc/core/arc_smp.c [hunk 3]
```diff
@@ -66,17 +66,41 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
...
 {
 	arch_cpustart_t fn;
 
 #ifdef CONFIG_SMP
+	struct arc_connect_bcr bcr;
+
+	bcr.val = z_arc_v2_aux_reg_read(_ARC_V2_CONNECT_BCR);
+
+	if (bcr.dbg) {
+	/* configure inter-core debug unit if available */
```
Comment (abrodkin) [inline arch/arc/core/arc_smp.c:94]: Either move this comment above `if` or add another identation level, i.e. move right to align with the next line `arc_connect_debug_mask_update(cpu_num);`.

File: arch/arc/core/arc_smp.c [hunk 4]
```diff
@@ -66,17 +66,41 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
...
+		arc_connect_debug_mask_update(cpu_num);
+	}
+
 	z_irq_setup();
 
 	z_arc_connect_ici_clear();
 	z_irq_priority_set(IRQ_ICI, ARCV2_ICI_IRQ_PRIORITY, 0);
 	irq_enable(IRQ_ICI);
+
```
Comment (abrodkin) [inline arch/arc/core/arc_smp.c:103]: Why adding this newline?

File: arch/arc/core/arc_smp.c [hunk 5]
```diff
@@ -117,6 +141,11 @@ static int arc_smp_init(const struct device *dev)
 
 	bcr.val = z_arc_v2_aux_reg_read(_ARC_V2_CONNECT_BCR);
 
+	if (bcr.dbg) {
+	/* configure inter-core debug unit if available */
```
Comment (abrodkin) [inline arch/arc/core/arc_smp.c:145]: Ditto

File: arch/arc/core/arc_smp.c (hunk 4 above)
Comment (IRISZZW) [reply arch/arc/core/arc_smp.c]: Removed.

File: arch/arc/core/arc_smp.c (hunk 3 above)
Comment (IRISZZW) [reply arch/arc/core/arc_smp.c]: Fixed.

File: arch/arc/core/arc_smp.c (hunk 5 above)
Comment (IRISZZW) [reply arch/arc/core/arc_smp.c]: Ditto.

File: arch/arc/core/arc_smp.c [hunk 6]
```diff
@@ -66,12 +66,35 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
...
 }
 
+#ifdef CONFIG_SMP
+static void arc_connect_debug_mask_update(int cpu_num)
+{
+	uint32_t core_mask = 1 << cpu_num;
+
+	core_mask |= z_arc_connect_debug_select_read();
+	z_arc_connect_debug_select_set(core_mask);
+	/* Debugger halt cores at conditions */
```
Comment (abrodkin) [inline arch/arc/core/arc_smp.c:76]: 1. Debugger halt**s**
2. Could you please list conditions while we're at it?

**Review by abrodkin — APPROVED** [review]

**Review by ruuddw — APPROVED** [review]

**Review by evgeniy-paltsev — APPROVED** [review]

**Pull request merged**